cgi-bin/DW/Controller/Shop.pm

```diff
@@ -124,4 +129,247 @@ sub shop_receipt_handler {
     return DW::Template->render_template( 'shop/receipt.tt', $vars );
 }
 
+# handles the shop checkout page
+sub shop_checkout_handler {
+    my ( $ok, $rv ) = _shop_controller( anonymous => 1 );
+    return $rv unless $ok;
+
+    my $cart  = $rv->{cart};
+    my $r     = DW::Request->get;
+    my $GET   = $r->get_args;
+    my $scope = 'shop/checkout.tt';
+
+    return error_ml("$scope.error.nocart") unless $cart;
+    return error_ml("$scope.error.emptycart") unless $cart->has_items;
+
+    # FIXME: if they have a $0 cart, we don't support that yet
+    return error_ml("$scope.error.zerocart")
+        if $cart->total_cash == 0.00 && $cart->total_points == 0;
+
+    # establish the engine they're trying to use
+    my $eng = DW::Shop::Engine->get( $GET->{method}, $cart );
+    return error_ml("$scope.error.invalidpaymentmethod")
+        unless $eng;
+
+    # set the payment method on the cart
+    $cart->paymentmethod( $GET->{method} );
+
+    # redirect to checkout url
+    my $url = $eng->checkout_url;
+    return $eng->errstr
+        unless $url;
+    return $r->redirect($url);
+
+}
+
+sub shop_history_handler {
+    my ( $ok, $rv ) = _shop_controller();
+    return $rv unless $ok;
+
+    my $cart   = $rv->{cart};
+    my $r      = DW::Request->get;
+    my $remote = $rv->{remote};
+
+    my @carts    = DW::Shop::Cart->get_all( $remote, finished => 1 );
+    my @cartrows = map { $_->{date} => DateTime->from_epoch( epoch => $_->starttime ) } @carts;
+
+    return DW::Template->render_template( 'shop/history.tt', { carts => \@carts } );
```

Comment:
Did you mean to use `@cartrows` here? You assign it in the previous line and then don't use it.